File: bindings/java/tests/SmokeTest.java
// SPDX-License-Identifier: 0BSD

package io.quad4.rrc;

import io.quad4.rrc.ffi.RrcLibrary;

public final class SmokeTest {
    public static void main(String[] args) {
        if (!Rrc.API_VERSION.equals(Rrc.version())) {
            fail("version mismatch: " + Rrc.version());
        }

        try (Node node = Node.create()) {
            node.start();
            node.stop();
        }

        try (Identity identity = Identity.generate()) {
            if (identity.hashBytes().length != Rrc.HASH_LEN) {
                fail("hash bytes length");
            }
        }

        byte[] sender = new byte[Rrc.HASH_LEN];
        for (int i = 0; i < sender.length; i++) {
            sender[i] = (byte) i;
        }
        try (Envelope env = Envelope.create(RrcLibrary.RRC_TYPE_MSG, sender)) {
            env.setRoom("lobby");
            env.setBodyText("hello");
            byte[] data = env.marshal();
            try (Envelope got = Envelope.unmarshal(data)) {
                if (!"hello".equals(got.bodyText())) {
                    fail("body mismatch");
                }
            }
        }

        if (!"#lobby".equals(Util.normalizeRoom("  #Lobby "))) {
            fail("normalize room");
        }
        if (!"alice".equals(Util.sanitizeNick(" alice "))) {
            fail("sanitize nick");
        }

        System.out.println("java smoke tests ok");
    }

    private static void fail(String msg) {
        System.err.println(msg);
        System.exit(1);
    }
}

File: bindings/kotlin/tests/SmokeTest.kt
// SPDX-License-Identifier: LicenseRef-Reticulum

package io.quad4.lxmf.kotlin

fun main() {
    check(LxmfKt.version() == LxmfKt.API_VERSION) { "version mismatch: ${LxmfKt.version()}" }

    LxmfIdentity.generate().use { identity ->
        val hash = identity.hashBytes()
        check(hash.size == LxmfKt.HASH_LEN) { "hash length" }
        check(identity.publicKey().isNotEmpty()) { "public key empty" }
        check(identity.deliveryHash().size == LxmfKt.HASH_LEN) { "delivery hash length" }

        LxmfMessage.create(hash, hash, "hi", "hello lxmf").use { msg ->
            val packed = msg.pack(identity)
            check(packed.isNotEmpty()) { "pack empty" }
            val inner = msg.encryptedPayload()
            check(inner.size == packed.size - LxmfKt.HASH_LEN) { "encrypted payload length" }
            LxmfMessage.unpack(packed).use { got ->
                check(got.content() == "hello lxmf") { "content mismatch: ${got.content()}" }
                check(got.title() == "hi") { "title mismatch: ${got.title()}" }
            }
        }
    }

    println("kotlin-lxmf smoke tests ok")
}
